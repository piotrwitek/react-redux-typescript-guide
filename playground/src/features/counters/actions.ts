import { action } from 'typesafe-actions';

import { ADD, INCREMENT } from './constants';

// CLASSIC API
export const increment = () => action(INCREMENT);
export const add = (amount: number) => action(ADD, amount);

// ALTERNATIVE API - allow to use reference to "action-creator" function instead of "type constant"
// e.g. case getType(increment): return { ... }
// This will allow to completely eliminate need for "constants" in your application, more info here:
// https://github.com/piotrwitek/typesafe-actions#constants

// OPTION 1 (with generics):
// import { createStandardAction } from 'typesafe-actions';
// export const increment = createStandardAction(INCREMENT)<void>();
// export const add = createStandardAction(ADD)<number>();

// OPTION 2 (with resolve callback):
// import { createAction } from 'typesafe-actions';
// export const increment = createAction(INCREMENT);
// export const add = createAction(ADD, resolve => {
//   return (amount: number) => resolve(amount);
// });
