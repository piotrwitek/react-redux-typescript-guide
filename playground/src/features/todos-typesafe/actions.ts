import cuid from 'cuid';
import { createAction } from 'typesafe-actions';

import { TodosFilter, Todo } from './models';

const ADD = 'todos/ADD';
const TOGGLE = 'todos/TOGGLE';
const CHANGE_FILTER = 'todos/CHANGE_FILTER';

export const add = createAction(ADD, (title: string) => ({
  id: cuid(),
  title,
}))<Todo>();

export const toggle = createAction(TOGGLE)<{ id: string }>();

export const changeFilter = createAction(CHANGE_FILTER)<TodosFilter>();
