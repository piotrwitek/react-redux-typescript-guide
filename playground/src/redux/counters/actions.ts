import { createAction } from 'typesafe-actions';

export const actions = {
  increment: createAction('INCREMENT'),
  add: createAction('ADD', (amount: number) => ({
    type: 'ADD',
    payload: amount,
  })),
};
