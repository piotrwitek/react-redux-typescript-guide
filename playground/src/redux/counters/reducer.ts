import { combineReducers } from 'redux';
import { getType } from 'typesafe-actions';

import { RootAction } from '@src/redux';

import { countersActions } from './';

export type CountersState = {
  readonly reduxCounter: number;
};

export const countersReducer = combineReducers<CountersState, RootAction>({
  reduxCounter: (state = 0, action) => {
    switch (action.type) {
      case getType(countersActions.increment):
        return state + 1; // action is type: { type: "INCREMENT"; }

      case getType(countersActions.add):
        return state + action.payload; // action is type: { type: "ADD"; payload: number; }

      default:
        return state;
    }
  },
});
