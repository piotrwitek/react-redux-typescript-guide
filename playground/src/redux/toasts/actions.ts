import { createAction } from 'typesafe-actions';

export type IToast = {
  id: string;
  text: string;
};

export const actions = {
  addToast: createAction('ADD_TOAST', (toast: IToast) => ({
    type: 'ADD_TOAST',
    payload: toast,
  })),
  removeToast: createAction('REMOVE_TOAST', (id: string) => ({
    type: 'REMOVE_TOAST',
    payload: id,
  })),
};
