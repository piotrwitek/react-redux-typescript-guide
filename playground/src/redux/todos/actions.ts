import { createAction } from 'typesafe-actions';
import cuid from 'cuid';

import { ITodosFilter } from './types';

const ADD_TODO = '@@todos/ADD_TODO';
const TOGGLE_TODO = '@@todos/TOGGLE_TODO';
const CHANGE_FILTER = '@@todos/CHANGE_FILTER';

export const addTodo = createAction(ADD_TODO,
  (title: string) => ({
    type: ADD_TODO,
    payload: {
      id: cuid(),
      title: title,
      completed: false,
    },
  })
);

export const toggleTodo = createAction(TOGGLE_TODO,
  (id: string) => ({ type: TOGGLE_TODO, payload: id })
);

export const changeFilter = createAction(CHANGE_FILTER,
  (filter: ITodosFilter) => ({ type: CHANGE_FILTER, payload: filter })
);
