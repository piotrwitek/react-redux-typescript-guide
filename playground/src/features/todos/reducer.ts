import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import { Todo, TodosFilter } from './models';
import * as actions from './actions';
import { ADD, CHANGE_FILTER, TOGGLE } from './constants';

export type TodosState = {
  readonly todos: Todo[];
  readonly todosFilter: TodosFilter;
};

export type TodosAction = ActionType<typeof actions>;

export default combineReducers<TodosState, TodosAction>({
  todos: (state = [], action) => {
    switch (action.type) {
      case ADD:
        return [...state, action.payload];

      case TOGGLE:
        return state.map(item =>
          item.id === action.payload
            ? { ...item, completed: !item.completed }
            : item
        );

      default:
        return state;
    }
  },
  todosFilter: (state = TodosFilter.All, action) => {
    switch (action.type) {
      case CHANGE_FILTER:
        return action.payload;

      default:
        return state;
    }
  },
});
