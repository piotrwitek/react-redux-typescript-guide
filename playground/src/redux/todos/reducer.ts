import { combineReducers } from 'redux';
import { getType } from 'typesafe-actions';

import { ITodo, ITodosFilter } from './types';
import * as todosActions from './actions';

export type TodosState = {
  readonly isFetching: boolean;
  readonly errorMessage: string | null;
  readonly todos: ITodo[];
  readonly todosFilter: ITodosFilter;
};

export type RootState = {
  todos: TodosState;
};

export const todosReducer = combineReducers<TodosState, TodosAction>({
  isFetching: (state = false, action) => {
    switch (action.type) {
      default:
        return state;
    }
  },
  errorMessage: (state = '', action) => {
    switch (action.type) {
      default:
        return state;
    }
  },
  todos: (state = [], action) => {
    switch (action.type) {
      case getType(todosActions.addTodo):
        return [...state, action.payload];

      case getType(todosActions.toggleTodo):
        return state.map((item) => item.id === action.payload
          ? { ...item, completed: !item.completed }
          : item
        );

      default:
        return state;
    }
  },
  todosFilter: (state = '', action) => {
    switch (action.type) {
      case getType(todosActions.changeFilter):
        return action.payload;

      default:
        return state;
    }
  },
});

import { $call } from 'utility-types';
const returnsOfActions = Object.values(todosActions).map($call);
export type TodosAction = typeof returnsOfActions[number];
