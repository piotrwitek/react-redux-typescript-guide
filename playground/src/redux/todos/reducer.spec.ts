import { todosReducer, TodosState, TodosAction } from './reducer';
import { addTodo, changeFilter, toggleTodo } from './actions';

/**
 * FIXTURES
 */
const activeTodo = { id: '1', completed: false, title: 'active todo' };
const completedTodo = { id: '2', completed: true, title: 'completed todo' };

const initialState = todosReducer(undefined, {});

/**
 * SCENARIOS
 */
describe('Todos Logic', () => {

  describe('initial state', () => {
    it('should match a snapshot', () => {
      expect(initialState).toMatchSnapshot();
    });
  });

  describe('adding todos', () => {
    it('should add a new todo as the first active element', () => {
      const action = addTodo('new todo');
      const state = todosReducer(initialState, action);
      expect(state.todos).toHaveLength(1);
      expect(state.todos[0].id).toEqual(action.payload.id);
    });
  });

  describe('toggling completion state', () => {
    it('should mark as complete todo with id "1"', () => {
      const action = toggleTodo(activeTodo.id);
      const state0 = { ...initialState, todos: [activeTodo] };
      expect(state0.todos[0].completed).toBeFalsy();
      const state1 = todosReducer(state0, action);
      expect(state1.todos[0].completed).toBeTruthy();
    });
  });

});
