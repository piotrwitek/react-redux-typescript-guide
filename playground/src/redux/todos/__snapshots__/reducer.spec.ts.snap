// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Logic initial state should match a snapshot 1`] = `
Object {
  "errorMessage": "",
  "isFetching": false,
  "todos": Array [],
  "todosFilter": "",
}
`;
